doc/compiler/optimizer/DataFlowEngine.md

```diff
@@ -0,0 +1,123 @@
+
+Reaching definitions is an example of a forward union set data-flow analysis while [liveness](https://github.com/eclipse/omr/blob/f2bc4f8f6eb09f6cc8fc4ba48717de4880b970e3/compiler/optimizer/DataFlowAnalysis.hpp#L840-L871) is an example of a backward union set data-flow analysis. 
+
+### Leaf Classes
+
+The final step in the virtual hierarchy are the leaf classes, which typically map to textbook examples of data-flow analyses. [`TR_ReachingDefinitions`](https://github.com/eclipse/omr/blob/f2bc4f8f6eb09f6cc8fc4ba48717de4880b970e3/compiler/optimizer/DataFlowAnalysis.hpp#L480-L504) and [`TR_Liveness`](https://github.com/eclipse/omr/blob/f2bc4f8f6eb09f6cc8fc4ba48717de4880b970e3/compiler/optimizer/DataFlowAnalysis.hpp#L840-L871) are two examples of leaf classes. By the time we reach the leaf classes, many of the virtual functions have already been implemented for us by the super classes. You may note that there isn't a lot of code in these leaf classes. This is a good thing, and it means the APIs in the virtual hierarchy have implemented most of the functionality for us, and we as developers are only left with modeling some specifics about our particular data-flow analysis, such as initialization and counting.
+
+## Operation on Structures and Regions
+
+As previously discussed, _gen_ and _kill_ sets are typically thought of in terms of basic blocks, however in the OMR data-flow engine they are thought of in terms of [`TR_Structure`](https://github.com/eclipse/omr/blob/f2bc4f8f6eb09f6cc8fc4ba48717de4880b970e3/compiler/optimizer/Structure.hpp#L58-L257). This design decision was made to give us the freedom to be able to generate _gen_ and _kill_ sets not just on a black basis but in terms of larger program fragments which improves performance of the OMR compiler when used in JIT compilation environments.
+
+A `TR_Structure` is composed one or more:
+
+- Basic blocks
+- `TR_Region`
+    - A well behaved loop for example which has one entry which dominates all the other regions in the loop
+- Acyclic structure
+    - Collection of structures where there is no cycle
+- Improper region
+    - Collection of structures where there _may_ be a cycle
+
+Computing the structure is a prerequisite to running the data-flow engine. Structures allow us to work on larger program fragments ("chunks") which improves compile time performance. Structures also have a caching mechanism built in for data-flow analysis in which we can keep track of the _in_ and _out_ sets while running a data-flow analysis. If we for example find the same _in_ set we can return the computed cached _out_ set since the program is not altered during data-flow analysis.
+
+Some data-flow analyses may get disinterested in using structures (and therefore _gen_ and _kill_ sets) if improper regions are encountered. This is true for the backwards analyses at the moment. They will fall back to tree level traversal rather than using _gen_ and _kill_ sets.
```

Comment:
I think this paragraph can be improved for accuracy. There are two different notions of using gen and kill sets. 

1. Using gen and kill sets for each block structure that means treetops do not need to be walked as often during the actual dataflow analysis (as described earlier in this document)
2. Using gen and kill sets for entire region structures at a time that means that the subnodes do not need to be traversed within the region structure

These two notions of using gen and kill sets are controlled by different queries in the dataflow analysis infrastructure. 

 The presence of an improper region only affects the latter notion (2) of using gen and kill sets on region structures. In other words, the presence of an improper region does not affect whether we will not walk the treetops, and that only depends on whether gen and kill sets are supported on even block structures or not.

Reply:
Fixed in [force-push](https://github.com/eclipse/omr/compare/8095a7e4089e396c82f174ef26070ae924586959..1cb9e0b5d0bf386498b82b1c4b566f62c74b8967). Thanks for elaborating!